README.rst
==============
OGLplus README
==============

:Author: Matúš Chochlík <chochlik@gmail.com>

.. contents::

.. _OpenGL: http://opengl.org/
.. _OpenAL: http://openal.org/
.. _EGL: http://www.khronos.org/egl
.. _OGLplus: http://oglplus.org/
.. _CMake: http://www.cmake.org/
.. _Doxygen: http://www.doxygen.org/
.. _Inkscape: http://inkscape.org/
.. _libPNG: http://www.libpng.org/
.. _GLEW: http://glew.sourceforge.net/
.. _GL3W: http://github.com/shakesoda/gl3w
.. _GLFW: http://www.glfw.org/
.. _FreeGLUT: http://freeglut.sourceforge.net/
.. _SDL: http://www.libsdl.org/
.. _wxGL: http://www.wxwidgets.org/
.. _Qt: http://qt.digia.com/

Introduction to OGLplus
=======================

`OGLplus`_ is collection of portable open-source libraries which implement thin
object-oriented facades over the `OpenGL`_  (version 3 and higher),
`OpenAL`_ (version 1.1) and `EGL`_ (version 1.4) C-language APIs.
It provides wrappers which automate resource and object management and make
the use of OpenGL, OpenAL and EGL in C++ safer and easier.



Building and Installation
=========================

For the impatient:
------------------

::

 # linux and similar *nix systems
 $> ./configure.sh
 # or
 $> ./configure.sh --prefix=/path/to/install

 $> cd _build
 $> make
 $> make install

 # windows
 $> .\configure.bat
 # open the generated MSVC solution in the _build directory
 # or use some variant of make if available

 # On platforms with python
 $> python configure.py --build

For the busy:
-------------

::

 # linux and similar *nix systems
 $> ./configure.sh [--prefix=/path/to/install] [--no-docs] --no-examples --build --install

Overview
--------

OGLplus uses a CMake-based build/configuration system. The library itself
is header-only, which means that applications using it do not need to link
to a compiled library, but need just to include the header files [#oglplus_link_library]_.

The build system handles several important tasks:

 * Detects if the necessary things are installed and makes a site-configuration
   header file

 * Detects the support for several C++11 features and builds a config header

 * Builds several additional, automatically generated headers

 * Installs all header files to a directory specified by the install prefix

 * Builds the example executables and assets (textures, models, etc.)
   used by the examples (optional)

 * Builds and installs the documentation (optional)


Requirements
------------

 - Compiler supporting required C++11 features [#req_cxx11_feats]_.
   Currently supported compilers:

   * ``g++`` (at least version 4.5, 4.6 and higher is recommended)

   * ``clang++`` (at least version 3.0) - possibly with some limitations
     due to the lack of support for some C++11 features

   * ``MSVC 2012`` - with some limitations due to the lack of support for
     some C++11 features


 - `CMake`_ (required)

 - `Doxygen`_ (optional) is required to build the documentation. This can be disabled
   with the ``--no-docs`` command line option of the ``configure`` script (see below).

 - `Inkscape`_ (optional) is used to convert textures for the examples from SVG to PNG.
   This is required only if the textures are not pre-built (typically
   when checked out from the repository, packaged releases are shipped with
   pre-built textures). Building of the textures is optional, they are not
   necessary when the building of examples is disabled.

 - A library defining the OpenGL API (required) -- the ``GL/glcorearb.h`` or ``GL3/gl3.h``
   headers or `GLEW`_, `GL3W`_, etc. and the corresponding binary library (``libGL.so``,
   ``OpenGL32.lib``, ``libGLEW.so``, etc.).  OGLplus does not define the OpenGL symbols
   (types, constants, functions, etc.), therefore applications using
   it need to define them themselves (before including OGLplus). The examples
   currently need GLEW (at least version 1.9) or the ``GL/glcorearb.h`` header
   (available for download from www.opengl.org/registry/api/glcorearb.h) and
   a GL binary library exporting the OpenGL (3 or higher) functions.
   The build system detects the presence of GLEW or ``GL/glcorearb.h`` and configures
   compilation and linking of the examples accordingly.
   Note, however, that if several options (like both GLEW and ``GL/glcorearb.h`` plus
   the binary GL lib) are available it may be necessary to specify which option to use.
   On Linux and similar systems the precedence is following: ``GL/glcorearb.h + libGL.so``,
   ``GL3/gl3.h + libGL.so``, `GLEW`_ and `GL3W`_ (the first one found is used, unless
   specified otherwise).  On Windows systems the precedence is:  `GLEW`_, `GL3W`_,
   ``GL/glcorearb.h + OpenGL.lib`` and ``GL3/gl3.h + OpenGL.lib``.
   Also note, that on systems with multiple versions of ``libGL.so`` (for example one
   provided by Mesa3D and another provided by your GPU vendor) it may be necessary
   to specify with the ``--library-dir`` option to the ``configure`` script (described
   below) in which directories to search for the library.
   The library to be used can be explicitly specified with the ``--use-gl-header-lib``
   option or with one of the ``--use-*`` options of the ``configure`` script.

 - A library initializing the default rendering context (required) -- Currently
   the examples can be built if at least one of the following libraries is
   installed on the system: X11+GLX, `FreeGLUT`_, `GLFW`_, `SDL`_, `wxGL`_ or `Qt`_.
   The build system detects the presence of these libraries and configures
   compilation and linking of the examples accordingly.
   The library to be used can be explicitly specified with the ``--use-gl-header-lib``
   option or with one of the ``--use-*`` options of the ``configure`` script (see below).


 - `libPNG`_ (optional) -- Some examples and some classes provided by OGLplus use libPNG to load
   PNG files. These are however not required for the general use of OGLplus,
   applications may use other means to load binary image files. The build system
   tries to detect the availability of libPNG and if not found the examples
   using it are not built.


On Linux distributions with the ``apt`` package manager, the following should
be enough to install most of the dependencies for the FreeGLUT+GLEW configuration:

::

 sudo apt-get install doxygen cmake g++ libglew-dev freeglut3-dev libpng12-dev

For the configuration using GLFW+GLEW you would need the following:

::

 sudo apt-get install doxygen cmake g++ libglew-dev libglfw-dev libpng12-dev

These two configs mentioned above are usually the ones that work on most systems.
Of course other combinations of the 'GL-API' and 'GL-Context' libraries
are supported and may be used as explained above.
For other configurations using SDL, Qt4, wxWidgets, etc. you need to install
the appropriate packages (the names vary wildly between distrubutions or even
between versions of the same distribution so they are not listed here).


CMake-based build configuration
-------------------------------

The CMake script defines and uses several variables to modify the build
configuration, which can be specified on the command-line when invoking
cmake (with the -D option. see cmake manual for details):

 * ``HEADER_SEARCH_PATHS`` *<empty>*: (semicolon-separated) list of paths
   to additional directories to search when looking for 3rd-party headers
   like GL/glew.h, GL3/gl3.h, GL/glcorearb.h, etc.

 * ``LIBRARY_SEARCH_PATHS`` *<empty>*: (semicolon-separated) list of paths
   to additional directories to search when looking for 3rd-party binary
   libraries like GL, GLEW, GL3W, GLFW, SDL, glut, png, etc.

 * ``OGLPLUS_NO_EXAMPLES`` *Off*: Do not build the examples nor the assets.

 * ``OGLPLUS_NO_DOCS`` *Off*: Do not build and install the documentation.


User-friendly configuration script
----------------------------------

The configuration script comes in three flawors:

1. ``configure.sh`` -- For platforms with bash.
2. ``configure.bat`` -- For windows.
3. ``configure.py`` -- For platforms with python (recommended).
                       This version of the script is the most portable
                       and supports most features.

The ``configure`` script is a more user-friendly way to invoke cmake and to specify
additional parameters for the configuration process.

Some of the more important command-line options are described below:

--help  Display the help screen.

--prefix PATH       Specifies the installation prefix path for cmake (sets
                    the value of the CMAKE_INSTALL_PREFIX variable).
                    If this option is not specified, cmake's default prefix is used.

--include-dir PATH    Specify additional directiories
                      to search when looking for header files. It may be used multiple
                      times to specify multiple directories. Headers are searched
                      in the directories specified with this option in the same order
                      in which they appear on the command-line and the default system
                      header locations are searched only afterwards. The first header
                      found is used, in case there are multiple versions of the searched
                      header file.


--library-dir PATH    Specify additional directiories
                      to search when looking for compiled libraries. It may be used
                      multiple times to specify multiple directories. Libraries are
                      searched in the directories specified with this option
                      in the same order in which they appear on the command-line
                      and the default system library locations are searched afterwards.
                      The first library found is used, in case there are multiple
                      versions of the searched library.

--no-docs  Do not build the documentation.

See the ``--help`` option for the full description and detailed info on the usage
of this script. Bash users can also do ``source config/configure.bash_complete``
to install a bash completion extension that allows to TAB-complete the command
line arguments for the ``configure-oglplus`` configuration script that invokes
``configure.py``.


Other build environments
========================

Microsoft Visual Studio 2012
----------------------------

OGLplus contains a solution file and a set of project files for building
some of the examples in MSVC 2012.
The solution is located at the following path:

- ``$(OGLplusRoot)/etc/msvc11/OGLplus/OGLplus.sln``

To build properly, it requires GLEW (1.9 or higher) and FreeGLUT libraries
to be installed and added to the list of header/library search paths
in the solution. This can be achieved by setting the following variables
visible to MSVC:

* ``OGLPLUS_DEP_INCLUDE_DIRS`` - list of (semicolon-separated) paths pointing
  to the directories where GLUT's and GLEW's header files are located.

* ``OGLPLUS_DEP_LIBRARY_DIRS`` - list of (semicolon-separated) paths pointing
  to the directories where GLUT's and GLEW's .lib files are located.

* ``OGLPLUS_DEP_LIBS`` - list of (semicolon-separated) .lib files to be linked
  to the examples. Typically ``glew32.lib;freeglut.lib`` for dynamic linking
  or ``glew32s.lib;freeglut_static.lib`` for static linking.

* ``OGLPLUS_PP_DEFS`` - list of (semicolon-separated) preprocessor symbols
  that should be defined. Typically ``GLEW_STATIC;FREEGLUT_STATIC`` are
  required for static linking of GLEW and FreeGLUT.

These variables can be set either as system or user environment variables
or the ``OGLplus-site.props`` property sheet file (located in the solution's
directory) can be used. Either edit the file in a text editor or use MSVC's
*Property Manager* and set the variables listed above as appropriate
on your system. Note, that the ``OGLplus-site.props`` file already
defines variables like the above, but with appended underscore
(for example ``OGLPLUS_DEP_INCLUDE_DIRS_``, etc.). To set the variables
you can open the file in a text editor, remove the underscores and set
the values to match your system configuration.

Also note that when building the examples via the MSVC solutions
the ``$(OGLplusRoot)/_prebuilt/`` directory is required. This directory
contains automatically generated textures and other files that are
not in the source repository. This directory is included in the release
packages. If you are building from sources from the Git repository
also download a recent package and copy the ``_prebuilt/`` directory
into to the directory where you have cloned the repo or use the cmake
build system and build them that way.


Assets used in examples
=======================

Many of the examples use binary media files which are not suitable
for inclusion to the source repository mainly because of their size.
OGLplus examples use texture bitmaps, fonts and models, OALplus examples use sounds,
etc. Some of these media files can be generated from textual representations
(for example SVG files converted or POVray sources rendered to bitmaps).
Assets for which there is a textual source and its size does not exceed
certain limits are included in the repository and are rendered into their
binary form by the build system. This may however require some third-party
applications and the rendering process can be very time-consuming.

Because of this, some of the medium-sized assets are pre-built and shipped
with the release packages (in the ``_prebuilt/`` directory) and the larger
assets can be downloaded individually from
http://sourceforge.net/projects/oglplus/files/assets/


Acknowledgements
================

- *Tolga Dalman* for helping to get started with CMake.
- *Timo Keller* for contributing one of the examples.
- *Denis Ovod* for constructive critique and suggestions.
- *Per Nordlöw* for useful suggestions.
- *Dmitry Yakimenko (detunized)* for several patches and porting to MacOS X.
- *Csala Tamás* for useful feedback and help with the fixing of several bugs.
- *David McFarland* for several patches.
- *Jesús Zazueta* for useful feedback and help with the fixing of several bugs.
- *Konstantin Gainullin* for porting to Marmalade and fixing several warnings.
- *Peter Laufenberg* for several patches.
- *Jan P Springer (regnirpsj)* for contributing SSBO wrappers and example.
License
=======

Copyright 2008-2015 Matus Chochlik. Distributed under the Boost
Software License, Version 1.0. (See accompanying file
LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)


.. [#oglplus_link_library] Although OGLplus is generally a header-only library,
   several parts of it (mostly some complex functions or functions with static data)
   can optionally be built separately and linked to applications, which can lead
   to improved build times especially for larger projects.
   See the documentation for the ``OGLPLUS_LINK_LIBRARY`` preprocessor configuration
   option for more details. Generally if ``OGLPLUS_LINK_LIBRARY`` is set to zero
   then everything is inlined, otherwise some functions are only declared, but not
   defined and the ``oglplus/lib.hpp`` header that contains the definition
   of all such functions must be included in one of the translation
   units that are linked into the final application.

.. [#req_cxx11_feats] OGLplus requires the following C++11 features:
   The ``type_traits`` and ``tuple`` librares, variadic preprocessor macros, r-value
   references and move-constructors.
   OGLplus also uses (but has workarounds for or disables certain components
   and/or functions if the features are not available):
   Strongly typed enumerations, variadic templates, initializer lists, lambdas,
   defaulted and deleted functions, function template default arguments,
   constexpr, noexcept, nullptr, explicit conversion operators, unicode literals
   and user-defined literals.


CHANGELOG
Version 0.67.0

- LOW_PROFILE versions of the SizeType and BigSizeType types have been added.
- Some problems with building in the LOW_PROFILE mode have been fixed.
- The GL version and extension detection in the build system has been improved.
- Other minor changes have been made to the build system.
- The Quickbook documentation has been updated.

Version 0.66.0

- The DeferredHandler class and the Outcome<T> template have been refactored.
- New overload of Texture::Wrap has been added.
- The ClientContext and its SettingStack helper classes have been updated.
- Libpng allocation checks have been added to the PNG image loader.
- The GL3W include path has been fixed.
- Minor fixes in the build system have been made.
- The Quickbook documentation has been updated.

Version 0.65.0

- The devel/testing implementation of EGL was updated to support multisampling.
- The GLX and EGL example harnesses were refactored and updated
  to use multisampling when the --samples option is specified.
- The ViewportIndex and DrawBufferIndex types were added.
- Context and ClientContext were updated to use the new types.
- A wrapper for NV_fragment_coverage_to_color was added.
- The PolygonMode enumeration was updated.
- Most of the examples were changed to use GLSL 120/130/140/150.
- Several bugs and problems were fixed.
- The build system was updated to detect the availability of C++ override.
- The Quickbook documentation was updated.


Version 0.64.0

- The DeferredHandler and the Outcome<T> classes which allow
  to defer and/or cancel calling of error handlers were added.
- New overloads of Shader::Compile, Shader::CompileInclude,
  Program::Link and Program::Validate and ProgramPipeline::Validate,
  which return Outcome<T> instead of handling errors eagerly were added.
- Some of the examples were changed to show the usage of the new features.
- Some of the examples were changed to use GLSL 120/130/140/150.
- Some of the older examples were removed.
- The Quickbook documentation was updated.

Version 0.63.0

- Some clang -Weverything compilation warnings were fixed.
- New overloads of CameraMatrix::Perspective functions were added.
- The internals of OALplus and EGLplus were refactored.
- Support for Shader-storage buffer objects was added.
- A bug in the Cloud and BrushedMetal image generators were fixed.
- A problem with the NV path command enumeration was fixed.


Version 0.62.0

- Some clang -Weverything compilation warnings were fixed.
- The SizeType class was updated.
- Other minor changes were made in the implementation details.
- A missing-identifier bug was fixed in images::Image.
- The Quickbook documentation was updated.


Version 0.61.0

- The GL version detection in the build system was updated
  to work better on MacOS.
- A preprocessor switch for enabling/disabling of Boost.Config was added.
- Wrappers for ARB_clip_control were added.
- Exceptions in OALplus were updated.
  - Error code enumerations were added.
  - ErrorAL was added.
  - The Code member function was added to ErrorAL, ErrorALC and ErrorALUT.
- Some compiler warnings were silenced.
- The SizeType and BigSizeType classes were added and some occurences
  of GLsizei and GLsizeiptr were replaced by them.
- The Quickbook documentation was updated.


Version 0.60.0

- The Boolean wrapper was updated, the operator ~ was added.
- The Object::HasInstance function was added.
- The native GL context wrappers were updated.
- The DSA Texture, Buffer and Framebuffer were partially refactored.
- The SingleFace enumeration was added.
- The FragmentData and FragmentDataSlot typedefs were added.
- A new standalone example was added.
- The oalplus::CurrentContext class was renamed to ContextMadeCurrent.
- Missing #ifdefs were added around specializations of EnumAssocType.
- The Quickbook documentation was updated.

Version 0.59.0

- Problems with move construction and move-assignment of objects
  were fixed.
- The Boolean wrappers were added.
- The Quickbook documentation was updated.

Version 0.58.0

- Noexcept specifiers were added to some functions.
- The oalplus::GetALName function was added.
- Small changes were made to oalplus::Listener and Device wrappers.
- Several assignments constructors were fixed to prevent problems with
  self assignment.
- The GL/AL object wrappers were refactored.
- Most of the standalone examples were backported to MSVC.
- Some MSVC warnings were fixed.
- Collisions between preprocessor macros and variable names were fixed.
- A problem that caused access violations in GL feature tests was fixed.
- The Quickbook documentation was updated.

Version 0.57.0

- The setting stack functionality in oglplus::ClientContext was reimplemented.
- Noexcept specifiers were added to some functions.
- The script for auto-generating of the enumeration headers was updated,
  now almost everything related to the enumerations is auto-generated.
- A bug in the definition of EnumValueRange was fixed.
- Minor problems in the GLSL shaders in the examples were fixed.
- The Quickbook documentation was updated.

Version 0.56.0

- The new state getter/setter functions were added to oglplus::Context.
- The oglplus::ClientContext class was updated, wrappers for other GL state
  variables and functions were added.
- A problem with the wxGL example harness on Windows was fixed.
- A missing include was added to the GLFW3 example harness.
- The GLM interoperability classes were updated to work with both older
  and newer versions of GLM.
- Minor fixes were made in the enumeration types.
- The automatically-generated enumeration headers were and the related
  tools were updated.
- The EnumAssocType and EnumAssocGLType metafunctions were added.
- The build system was updated, check of support for strongly-typed
  enum template parameters was added.
- The quickbook documentation was updated.

Version 0.55.0

- The EnumToClass helper template and its specializations were added.

web/index.php
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta name="author" content="Matus Chochlik"/>
<meta name="copyright" content="Copyright 2010-2015 Matus Chochlik"/>
<meta name="keywords" content="C++ OpenGL Graphics Library"/>
<meta name="description" content="C++ wrapper for the OpenGL&reg; C API"/>

<title>OGLplus - a C++ wrapper for modern OpenGL®.</title>

<link rel="icon" type="image/x-icon" href="favicon.ico"/>
<link rel="shortcut icon" type="image/x-icon" href="favicon.ico"/>

<link rel="stylesheet/less" type="text/css" href="oglplus.less"/>
<script src="less.js" type="text/javascript"></script>


<link href="http://oglplus.org/news.rss" rel="alternate" type="application/rss+xml" title="OGLplus news feed" />

<script type="text/javascript">
(function() {
    var po = document.createElement('script'); po.type = 'text/javascript'; po.async = true;
    po.src = 'https://apis.google.com/js/plusone.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(po, s);
})();
</script>
<link href="https://plus.google.com/+OglplusOrg/" rel="publisher" />
</head>
<body>
<div id="main-title">
<h1>OGLplus: a C++ wrapper for modern OpenGL&copy;</h1>
</div>
<div id="frame">
	<div id="head">
		<div id="top">
				Visit the <em>OGLplus</em> pages on
				<a href="https://plus.google.com/+OglplusOrg/?prsrc=3" style="text-decoration: none;">
					<img src="https://ssl.gstatic.com/images/icons/gplus-32.png"
						width="32" height="32"
						style="border: 0;"
						alt="OGLplus page on Google+"
					/>
				</a>
				<a href="https://github.com/matus-chochlik/oglplus" style="text-decoration: none;">
					<img src="http://github.com/favicon.ico"
						style="border: 0;"
						alt="OGLplus on GitHub"
					/>
				</a>
				<a href="http://www.linkedin.com/groups/OGLplus-4592626" style="text-decoration: none;">
					<img src="http://www.linkedin.com/img/webpromo/btn_in_20x15.png"
						width="20" height="15"
						style="border: 0;"
						alt="OGLplus group on LinkedIn"
					/>
				</a>

				for recent news and more info.
				<hr class="top-div"/>
				<!-- AddThis Button BEGIN -->
				<div id="addthis-box" class="addthis_toolbox addthis_default_style ">
				<a class="addthis_button_facebook_like" fb:like:layout="button_count"></a>
				<a class="addthis_button_tweet"></a>
				<a class="addthis_button_google_plusone" g:plusone:size="medium"></a>
				<a class="addthis_counter addthis_pill_style"></a>
				</div>
				<script type="text/javascript" src="http://s7.addthis.com/js/250/addthis_widget.js#pubid=xa-4f35537052d88ff3"></script>
				<!-- AddThis Button END -->
		</div>
		<div id="logo">
			<img src="oglplus/html/oglplus.png"/>
		</div>
	</div>

	<div id="content">
		<div id="main">
			<div class="content-section">
				<h3>Introduction</h3>
				<p><em>OGLplus</em> is a collection of open-source, cross-platform libraries which implement an object-oriented facade over the OpenGL® (version 3 and higher) and also OpenAL® (version 1.1) and EGL (version 1.4) C-language APIs. It provides wrappers which automate resource and object management and make the use of these libraries in C++ safer and more convenient. <a href="oglplus/html/index.html">Read more...</a></p>
			</div>

			<div class="content-section">
				<div id="quick-links" class="content-column-1">
					<h3>Download</h3>
					<ul>
						<li id="direct-download-stable"><a href="http://sourceforge.net/projects/oglplus/files/latest/download"><b>Direct download</b></a> (latest stable)</li>
						<li id="direct-download-devel"><a href="http://github.com/matus-chochlik/oglplus/archive/develop.zip">Direct download</a> (development)</li>
						<li id="download"><a href="http://sourceforge.net/projects/oglplus/files/">All releases</a></li>
						<li id="assets"><a href="http://sourceforge.net/projects/oglplus/files/assets/">Assets (models, textures) used in examples</a></li>

						<li id="git-repo"><a href="https://github.com/matus-chochlik/oglplus">GIT repository</a></li>
					</ul>
					<h3>Documentation</h3>
					<ul>
						<li id="readme"><a href="http://github.com/matus-chochlik/oglplus#oglplus-readme">README</a></li>
						<li id="quickbook-doc"><a href="http://matus-chochlik.github.io/oglplus/doc/">Quickbook</a></li>
						<li id="doxygen-doc">Doxygen: <a href="oglplus/html/index.html">OGLplus</a>, <a href="oalplus/html/index.html">OALplus</a>, <a href="eglplus/html/index.html">EGLplus</a></li>
					</ul>
					<h3>Components</h3>
					<ul>
						<li><b>OGLplus</b> - a wrapper for modern OpenGL
						<ul>
							<li id="oglplus-examples"><a href="oglplus/html/examples.html">Examples</a></li>
							<li id="oglplus-tutorials"><a href="oglplus/html/oglplus_tutorials.html">Tutorials</a></li>
						</ul>
						</li>
						<li><b>OALplus</b> - a wrapper for OpenAL
						<ul>
							<li id="oalplus-examples"><a href="oalplus/html/examples.html">Examples</a></li>
							<li id="oalplus-tutorials"><a href="oalplus/html/oalplus_tutorials.html">Tutorials</a></li>
						</ul>
						</li>
						<li><b>EGLplus</b> - a wrapper for EGL
						<ul>
							<li id="eglplus-examples"><a href="eglplus/html/examples.html">Examples</a></li>
						</ul>
						</li>
					</ul>
				</div>

				<div id="features" class="content-column-2">
					<h3>Features</h3>
					<ul>
						<li><a href="oglplus/html/index.html#feat_auto_resource_management">Automatic resource management</a></li>
						<li><a href="oglplus/html/index.html#feat_type_safety">Type safety</a></li>
						<li><a href="oglplus/html/index.html#feat_error_handling">Error handling and diagnostic</a></li>
						<li><a href="oglplus/html/index.html#feat_interoperability_with_opengl">Interoperability with OpenGL C API</a></li>
						<li><a href="oglplus/html/index.html#feat_related_classes">Additional CG-related classes</a></li>
						<li><a href="oglplus/html/index.html#oglplus_object_description">Optional description of objects and resources</a></li>
						<li><a href="oglplus/html/index.html#oglplus_configurability">Configurability</a></li>
						<li>Utilities
						<ul>
							<li>Text rendering utilities</li>
							<li>Mesh loaders / generators</li>
							<li>Image loaders / generators</li>
						</ul>
						</li>
					</ul>
					<h3>Follow OGLplus</h3>
					<ul>
						<li id="github"><a href="https://github.com/matus-chochlik/oglplus">GitHub</a></li>
						<li id="sourceforge"><a href="http://sourceforge.net/projects/oglplus/">Sourceforge</a></li>
						<li id="g-plus"><a href="https://plus.google.com/+OglplusOrg/?prsrc=3">Google+ news page</a></li>
						<li id="linkedin"><a href="http://www.linkedin.com/groups/OGLplus-4592626">LinkedIn group</a></li>
						<li id="openhub"><a href="http://www.openhub.net/p/oglplus">OpenHub</a></li>
					</ul>
				</div>
			</div>

			<div class="content-section">
			<div id="screenshot-gallery" class="content-column-3">
				<script src="http://ajax.googleapis.com/ajax/libs/jquery/1/jquery.js"></script>
				<script src="galleria/galleria-1.2.5.min.js"></script>
				<div id="gallery">
					<img src="oglplus/html/007_voronoi.png"/>
					<img src="oglplus/html/021_translucent_arrow.png"/>
					<img src="oglplus/html/022_blender_mesh.png"/>
					<img src="oglplus/html/022_parallax_map.png"/>
					<img src="oglplus/html/023_sky.png"/>
					<img src="oglplus/html/026_furry_torus.png"/>
					<img src="oglplus/html/026_ssao.png"/>
					<img src="oglplus/html/026_stencil_shadow.png"/>
					<img src="oglplus/html/027_brain_scan_vis.png"/>
					<img src="oglplus/html/028_depth_map_sss.png"/>
					<img src="oglplus/html/028_ripples.png"/>
					<img src="oglplus/html/029_gpu_sort_tfb.png"/>
					<img src="oglplus/html/029_fire_hydrant.png"/>
					<img src="oglplus/html/030_light_rays.png"/>
					<img src="oglplus/html/030_pin_display.png"/>
					<img src="oglplus/html/031_brick_torus.png"/>
					<img src="oglplus/html/031_sketch.png"/>
					<img src="oglplus/html/032_bar_grid.png"/>
					<img src="oglplus/html/033_metal_and_glass.png"/>
					<img src="oglplus/html/033_steam.png"/>
					<img src="oglplus/html/040_jelly_cube.png"/>
				</div>
				<script>
					Galleria.loadTheme("galleria/themes/classic/galleria.classic.min.js");
					$("#gallery").galleria({
						width: 400,
						height: 400,
						autoplay: 2000
					});
				</script>
			</div>

			<div id=other" class="content-column-4">
				<h3>Media</h3>
				<ul>
					<li id="screenshots"><a href="oglplus/html/oglplus_screenshot_gallery.html">Screenshots</a></li>
					<li id="videos"><a href="https://plus.google.com/u/0/+OglplusOrg/videos">Videos</a></li>
				</ul>
				<h3>Related projects</h3>
				<ul>
					<li id="eagine"><a href="https://github.com/matus-chochlik/eagine">EAGine</a></li>
					<li id="angles"><a href="https://github.com/matus-chochlik/angle">Angles</a></li>
				</ul>
				<h3>Other resources</h3>
				<ul>
					<li><a href="https://github.com/jjzazuet/jgl" title="Java wrapper for OpenGL 3 and higher by Jesús Zazueta">JGL</a> by Jesús Zazueta</li>
					<li><a href="http://science.webhostinggeeks.com/c-omot-za-opengl" title="by Anja Skrba at Webhostinggeeks.com">Serbo-Croatian translation</a> by Anja Skrba</li>
					<li><a href="http://onlygraphix.com/2014/01/18/oglplus-tutorialdeferred-renderer/" title="by Michael Ivanov at http://onlygraphix.com/">Tutorial: Deferred renderer</a> by Michael Ivanov</li>
				</ul>
			</div>
			</div>

		</div>
		<div id="right">
			<?php
				$xml = new DOMDocument();
				$xml->load('news.rss');

				$xsl = new DOMDocument;
				$xsl->load('rss2html.xsl');

				$proc = new XSLTProcessor();
				$proc->importStyleSheet($xsl);
				$proc->setParameter('', 'feed-name', 'news.rss');

				echo $proc->transformToXML($xml);
			?>
		</div>
	</div>
	<div id="footer">
		<em>Copyright &copy; 2008-2015</em> <strong>Matúš Chochlík</strong>, Žilina, Slovakia.<br/>
		&lt;<a href="mailto:chochlik@gmail.com">chochlik -at -gmail.com</a>&gt;
		&lt;<a href="mailto:matus.chochlik@fri.uniza.sk">matus.chochlik -at- fri.uniza.sk</a>&gt;<br/>

		<p>Distributed under the Boost Software License, Version 1.0. (See a copy at
		<a href="http://www.boost.org/LICENSE_1_0.txt">http://www.boost.org/LICENSE_1_0.txt</a>)</p>
	</div>
</div>
</body>
</html>

